src/gui/editorwidgets/core/qgseditorwidgetautoconf.h

```diff
@@ -0,0 +1,83 @@
+/***************************************************************************
+    qgseditorwidgetautoconf.h
+    ---------------------
+    begin                : July 2016
+    copyright            : (C) 2016 by Patrick Valsecchi
+    email                : patrick.valsecchi at camptocamp.com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSEDITORWIDGETAUTOCONF_H
+#define QGSEDITORWIDGETAUTOCONF_H
+
+#include <QList>
+#include <QSharedPointer>
+
+class QgsVectorLayer;
+class QgsWidgetSetup;
+
+/** \ingroup gui
+ * Base class for plugins allowing to pick automatically a widget type for editing fields.
+ *
+ * @note added in QGIS 3.0
+ */
+class GUI_EXPORT QgsEditorWidgetAutoConfPlugin
+{
+  public:
+    /**
+     * Typical returned values are:
+     *   * 0: no matching type found.
+     *   * 10: a widget has been guessed from the type of field.
+     *   * 20: a widget has been determined from an external configuration (for example a database table)
+     *
+     * @param vl        The vector layer for which this widget will be created
+     * @param fieldName  The field name on the specified layer for which this widget will be created
+     * @param setup     Where the actual setup is returned
+     *
+     * @return and integer value rating how good is the setup provided by this plugin.
+     */
+    virtual unsigned int getSetup( const QgsVectorLayer& vl, const QString& fieldName, QgsWidgetSetup& setup ) = 0;
+};
+
+
+/** \ingroup gui
+ * Singleton that allows to register plugins to pick automatically a widget type for editing fields.
```

Comment:
Would it make sense to merge this into `QgsEditorWidgetRegistry`? As far as I can see, it's initialized and managed in there. It will be easier to cleanup and manage garbage collection if things are not global static.